pkg/cmd/pulumi/state_edit.go

```diff
@@ -0,0 +1,269 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+
+	survey "github.com/AlecAivazis/survey/v2"
+	surveycore "github.com/AlecAivazis/survey/v2/core"
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/diag/colors"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/result"
+	"github.com/spf13/cobra"
+)
+
+func newStateEditCommand() *cobra.Command {
+	var stackName string
+	stateEdit := &stateEditCmd{
+		Ctx:       commandContext(),
+		Colorizer: cmdutil.GetGlobalColorization(),
+	}
+	cmd := &cobra.Command{
+		Use: "edit",
+		// TODO(dixler) Add YAML + comments with sorted keys before exposing this to users.
+		Hidden: !hasExperimentalCommands(),
+		Short:  "Edit the current stack's state in your EDITOR",
+		Long: `[EXPERIMENTAL] Edit the current stack's state in your EDITOR
+
+This command can be used to surgically edit a stack's state in the editor
+specified by the EDITOR environment variable and will provide the user with
+a preview showing a diff of the altered state.`,
+		Args: cmdutil.NoArgs,
+		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {
+			if !cmdutil.Interactive() {
+				return result.Error("pulumi state edit must be run in interactive mode")
+			}
+			s, err := requireStack(commandContext(), stackName, stackLoadOnly, display.Options{
+				IsInteractive: true,
+			})
+			if err != nil {
+				return result.FromError(err)
+			}
+			if err := stateEdit.Run(s); err != nil {
+				return result.FromError(err)
+			}
+			return nil
+		}),
+	}
+	cmd.PersistentFlags().StringVar(
+		&stackName, "stack", "",
+		"Remove the stack and its config file after all resources in the stack have been deleted")
+	return cmd
+}
+
+type stateEditCmd struct {
+	Stdin     io.Reader
+	Stdout    io.Writer
+	Colorizer colors.Colorization
+	Ctx       context.Context
+}
+
+type snapshotBuffer struct {
+	Name     func() string
+	Snapshot func() (*deploy.Snapshot, error)
+	Reset    func() error
+	Cleanup  func()
+}
+
+func newSnapshotBuffer(fileExt string, sf snapshotEncoder, snap *deploy.Snapshot) (*snapshotBuffer, error) {
```

Comment:
This is implementing a class manually, using the closure to hold the actual state.
Any reason not to do this:

```go
type snapshotBuffer struct {
  path string
  originalText []byte
}

func newSnapshotBuffer(...) (*snapshotBuffer, error) {
  // ...
  tempFile.Close()

  // ..
  sb := &snapshotBuffer{path: tempFile.Name(), originalText: ...}
  // ...
}

func (sb *snapshotBuffer) Snapshot() (*deploy.Snapshot, error) { ... }

func (sb *snapshotBuffer) Reset() error {
  return os.WriteFile(sb.path, sb.originalText, 0600)
}

func (sb *snapshotBuffer) Cleanup() {
  _ = os.Remove(sb.path)
}
```